app/grandchallenge/discussion_forums/views.py

```diff
@@ -108,3 +120,116 @@ def get_success_url(self):
             "discussion-forums:topic-list",
             kwargs={"challenge_short_name": self.request.challenge.short_name},
         )
+
+
+class ForumPostCreate(ObjectPermissionRequiredMixin, CreateView):
+    model = ForumPost
+    permission_required = "create_topic_post"
+    raise_exception = True
+    form_class = ForumPostForm
+
+    @cached_property
+    def forum(self):
+        return self.request.challenge.discussion_forum
+
+    @cached_property
+    def topic(self):
+        return get_object_or_404(
+            ForumTopic, forum=self.forum, slug=self.kwargs["slug"]
+        )
+
+    def get_permission_object(self):
+        return self.topic
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data()
+        context.update({"forum": self.forum, "topic": self.topic})
+        return context
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs.update(
+            {"user": self.request.user, "topic": self.topic, "instance": None}
+        )
+        return kwargs
+
+    def get_success_url(self):
+        return f"{self.topic.get_absolute_url()}#post-{self.object.pk}"
```

Comment:
You may need to go to the last page here to ensure that the post is on the page. Same in `ForumPostUpdate`: this could all go in `get_absolute_url` for the post, no need to duplicate it.

Reply:
Ah right, with the pagination added, the page also neeeds to be taken into account when getting and displaying the form. At the moment I just append to the bottom of the post list, but that may not be the end of the list. I think I might need to not use HTMX for the create and update views after all. Looking into it now.

Reply:
If I switch to just a regular view that only shows the post form, the downside is that the user no longer sees the messages they're responding to. This was the reason why I went with HTMX in the first place. I guess the best is then to mimic the post list view and have the form view show the last page of posts for the topic with the form at the bottom.

Reply:
Mimicking the post list on the post create page is messy and requires quite a bit of logic to work properly (with the paginator especially), so I don't think that's the way to go. Is it an ok assumption to say that people will already be on the last page when replying, and hence just do what I do now and append the form to the bottom of the page. The redirect on save would still load the appropriate page and the new post would still appear as the last post. Having had a look through GC I have yet to find a topic with more than 10 posts, so we might never hit the pagination for the post list anyway.

Reply:
Especially if we also implement the last read functionality discussed on the pitch, a topic view would always open on the last page for the user. So I think it is fine to just append the form to the end of the current page.

Reply:
MIDOG has an announcement with 18 replies in case you wanted to see how pagination is currently implemented: https://grand-challenge.org/forums/forum/midog2021-598/

Reply:
Thanks for finding that. At the moment, the post create page shows the form to create a post and then dumps the most recent 10 posts underneath that form in a scrollable card. That's not very convenient, but users do have a way to see the latest posts they are replying to if they scroll down and find the that section. 

I also see in one of the comments that the organizer missed some posts which appeared on the second page because she wasn't aware of the second page. I guess the "open topic at last unread post" would fix that issue.

Reply:
Alright, I have bitten in the sour apple and turned both the PostCreate and PostUpdate posts into normal Django views. The create view now displays the latest 5 posts (seems sufficient to me) before the form, the update view does not (no need IMHO). 

I'm not 100% convinced that this a lot better than what I had before in terms of user friendliness, but it is slightly simpler from the developer perspective and hence easier to maintain.

Reply:
I thought that you were going to have the post create form at the bottom of the post list view? You can have a post a reply button and reveal it https://getbootstrap.com/docs/4.0/components/collapse/

Reply:
That is what I had (implemented with HTMX), but I couldn't figure out how that would work with pagination if the user is on the first page.

Reply:
I don't get it? Why not:

```
{% for post in page %}
  # render post
{% endfor %}

{% if last_page %}
  # Render new post form
{% else %}
  # Reply Button that links to last page with the anchor for the new post form
{% endif %}
```

Or always render it?

Reply:
I don't know. It seemed strange to me to pass in a form to a list view context. I haven't seen that done anywhere in the codebase, so then standard Django views seemed cleaner. What you suggests would make for the most user-friendly design though. I will continue working on this tomorrow.

Reply:
Can't quickly find a list view but we include forms in other views in several places, e.g. see `get_conversation_form` of `EvaluationDetail` or the user groups forms of `ReaderStudyDetail`, `AlgorithmDetail`, `JobDetail` etc.

Reply:
I tried this out and it works well. The only weird thing maybe is that if you edit the last post in a topic, you end up with an edit form (rendered into place via HTMX) just above a create form (which is always there after the last post). They have clear titles, and the user only sees the other form if they scroll down, so probably fine:

![image](https://github.com/user-attachments/assets/7c8464ba-712a-44d6-a042-7828a931310c)